core/primitives/src/upgrade_schedule.rs

```diff
@@ -0,0 +1,143 @@
+use chrono::{DateTime, NaiveDateTime, ParseResult, Utc};
+use near_primitives_core::types::ProtocolVersion;
+
+pub(crate) struct ProtocolUpgradeVotingSchedule(pub &'static str);
+impl ProtocolUpgradeVotingSchedule {
+    #[allow(dead_code)]
+    pub fn is_valid(self) -> bool {
+        self.parse().is_ok()
+    }
+
+    pub fn get(self) -> DateTime<Utc> {
+        DateTime::<Utc>::from_utc(self.parse().unwrap(), Utc)
+    }
+
+    fn parse(self) -> ParseResult<NaiveDateTime> {
+        NaiveDateTime::parse_from_str(&self.0, "%Y-%m-%d %H:%M:%S")
+    }
+}
+
+pub(crate) fn get_protocol_version_internal(
+    next_epoch_protocol_version: ProtocolVersion,
+    client_protocol_version: ProtocolVersion,
+    schedule: Option<ProtocolUpgradeVotingSchedule>,
+) -> ProtocolVersion {
+    if next_epoch_protocol_version >= client_protocol_version {
+        client_protocol_version
+    } else if let Some(voting_start) = schedule {
+        if chrono::Utc::now() < voting_start.get() {
```

Comment:
`chorono::Utc::now` potentially can go backwards, in effect a node can change its mind whether it supports the new protocol version or not. A somewhat more robust approach in theory would be to persist the decision to upgrade to the database. Or, perhaps even better, to base the decision on the the timestamp of the last block.